src/main/java/io/fabric8/jenkins/openshiftsync/SecretWatcher.java

```diff
@@ -0,0 +1,167 @@
+            onInitialSecrets(secrets);
+            logger.fine("handled Secrets resources");
+            if (secretWatch == null) {
+              secretWatch = getOpenShiftClient().secrets().inNamespace(namespace).withLabel(LABEL_JENKINS, VALUE_SYNC).withResourceVersion(secrets.getMetadata().getResourceVersion()).watch(SecretWatcher.this);
+            }
+          } catch (Exception e) {
+            logger.log(SEVERE, "Failed to load Secrets: " + e, e);
+          }
+        }
+
+      }
+    };
+    relister = Timer.get().scheduleAtFixedRate(task, 100, 10 * 1000, TimeUnit.MILLISECONDS);
+  }
+
+  public synchronized void stop() {
+    if (relister != null && !relister.isDone()) {
+      relister.cancel(true);
+      relister = null;
+    }
+    if (secretWatch != null) {
+      secretWatch.close();
+      secretWatch = null;
+    }
+  }
+
+  @Override
+  public synchronized void onClose(KubernetesClientException e) {
+    if (e != null) {
+      logger.warning(e.toString());
+
+      if (e.getStatus() != null && e.getStatus().getCode() == HTTP_GONE) {
+        stop();
+        start();
+      }
+    }
+  }
+
+  private synchronized void onInitialSecrets(SecretList secrets) {
+    List<Secret> items = secrets.getItems();
+    if (items != null) {
+      for (Secret secret : items) {
+        try {
+          upsertCredential(secret);
+        } catch (Exception e) {
+          logger.log(SEVERE, "Failed to update job", e);
+        }
+      }
+    }
+  }
+
+  @SuppressFBWarnings("SF_SWITCH_NO_DEFAULT")
+  @Override
+  public synchronized void eventReceived(Action action, Secret secret) {
+    try {
+      switch (action) {
+        case ADDED:
+          upsertCredential(secret);
+          break;
+        case DELETED:
+          deleteCredential(secret);
+          break;
+        case MODIFIED:
+          modifyCredential(secret);
+          break;
+      }
+    } catch (Exception e) {
+      logger.log(Level.WARNING, "Caught: " + e, e);
+    }
+  }
+
+  private void upsertCredential(final Secret secret) throws Exception {
+    if (isJenkinsSecret(secret)) {
+      CredentialsUtils.upsertCredential(secret);
+    }
+  }
+
+  private void modifyCredential(Secret secret) throws Exception {
+    if (isJenkinsSecret(secret)) {
+      upsertCredential(secret);
+      return;
+    }
+
+    // no longer a Jenkins build so lets delete it if it exists
+    deleteCredential(secret);
+  }
+
+  private static boolean isJenkinsSecret(Secret secret) {
+    ObjectMeta metadata = secret.getMetadata();
+    if (metadata != null) {
+      String name = metadata.getName();
+      return name != null && !name.isEmpty();
+    }
+    return false;
+  }
+
+  private void deleteCredential(final Secret secret) throws Exception {
+    // TODO should we delete credentials?
+    // if we do then we should update BuildConfigWatcher to do the same
```

Comment:
I think so we should create a new issue for the same and track in separate PR, So that this PR does not get more big. 
I can take delete credentials thing sometime later.